test/test_corpus.cc

```diff
@@ -601,6 +601,118 @@ TEST_P(LSPTest, All) {
+
+                    EXPECT_FALSE(get_if<JSONNullObject>(&receivedCodeActionResponse));
+
+                    if (!get_if<JSONNullObject>(&receivedCodeActionResponse)) {
+                        UnorderedMap<string, unique_ptr<CodeAction>> receivedCodeActionsByTitle;
+                        auto &receivedCodeActions = get<vector<unique_ptr<CodeAction>>>(receivedCodeActionResponse);
+                        for (auto &codeAction : receivedCodeActions) {
+                            bool codeActionTitleUnique =
+                                receivedCodeActionsByTitle.find(codeAction->title) == receivedCodeActionsByTitle.end();
+                            EXPECT_TRUE(codeActionTitleUnique)
+                                << "Found code action with duplicate title: " << codeAction->title;
+
+                            if (codeActionTitleUnique) {
+                                receivedCodeActionsByTitle[codeAction->title] = move(codeAction);
+                            }
+                        }
+
+                        vector<shared_ptr<ApplyCodeActionAssertion>> matchedCodeActionAssertions;
+
+                        // Check any code action assertions for the same error range, i.e. verify that when applied,
+                        // they match the expectation.
+                        auto it = applyCodeActionAssertions.begin();
+                        while (it != applyCodeActionAssertions.end()) {
+                            auto codeActionAssertion = it->get();
+                            // TODO(sushain): what if there are multiple errors on the same range?
+                            if (cmpRanges(*codeActionAssertion->range, *error->range) == 0) {
+                                auto it2 = receivedCodeActionsByTitle.find(codeActionAssertion->title);
+                                EXPECT_NE(it2, receivedCodeActionsByTitle.end()) << fmt::format(
+                                    "Did not receive code action matching assertion `{}` for error `{}`...",
+                                    codeActionAssertion->toString(), error->toString());
+                                if (it2 != receivedCodeActionsByTitle.end()) {
+                                    auto codeAction = move(it2->second);
+                                    codeActionAssertion->check(test.sourceFileContents, codeAction, test.testName,
+                                                               fileUri);
+                                    matchedCodeActionAssertions.emplace_back(*it);
+                                    it = applyCodeActionAssertions.erase(it);
+                                    continue;
+                                }
+                            }
+
+                            ++it;
+                        }
+
+                        if (exhaustiveApplyCodeAction) {
+                            if (matchedCodeActionAssertions.size() > receivedCodeActionsByTitle.size()) {
```

Comment:
I think both of these errors print out all code actions/action assertions since you don't remove the ones that matched right now.